From eb037cadac52581027c28f59c7bbab68b235ea12 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Wed, 16 Apr 2025 19:57:02 +0200
Subject: [PATCH] HACK: Block Genshin Impact from connecting on first request.

Fixes issue where Genshin refuses to launch if connected to the internet.
Blocking the first request is enough to make it work again. xd
---
 dlls/ws2_32/socket.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index 3a412b00c1e..60eccc3937d 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -1243,6 +1243,23 @@ int WINAPI closesocket( SOCKET s )
     return 0;
 }
 
+// HACK: Detect whether Genshin is running
+static BOOL is_genshin(void)
+{
+    static volatile char cache = -1;
+    BOOL ret = cache;
+    if (ret == -1)
+    {
+        const WCHAR *p, *name = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
+        if ((p = wcsrchr(name, '/')))
+            name = p + 1;
+        if ((p = wcsrchr(name, '\\')))
+            name = p + 1;
+        ret = !wcsicmp(name, L"GenshinImpact.exe");
+        cache = ret;
+    }
+    return ret;
+}
 
 /***********************************************************************
  *      connect   (ws2_32.4)
@@ -1254,6 +1271,17 @@ int WINAPI connect( SOCKET s, const struct sockaddr *addr, int len )
     HANDLE sync_event;
     NTSTATUS status;
 
+    // HACK: Block Genshin Impact from connecting to internet on first request.
+    // Fixes game not launching.
+    static int delay_start = 0;
+
+    if (!delay_start && is_genshin()) 
+    {
+        delay_start++;
+        FIXME("Genshin Impact detected, blocking connection..\n");
+        return SOCKET_ERROR;
+    }
+
     TRACE( "socket %#Ix, addr %s, len %d\n", s, debugstr_sockaddr(addr), len );
 
     if (!(sync_event = get_sync_event())) return -1;
-- 
2.49.0

